### PR TITLE
Fixed `move_and_slide` `stop_on_slope`

### PR DESCRIPTION
**Pull request by AndreaCatania**
After the `move_and_slide` behaviour reversion (in #33864) the old stop on slope check is not more valid making the `stop_on_slope` not working. This PR fix it for both 2d and 3d.

This work has been kindly sponsored by IMVU.

### CODE REVIEW

**Comment by akien-mga** [comment]
CC @madmiraal

**Comment by madmiraal** [comment]
@AndreaCatania, please explain what problem this change is solving, because, from what I can see, this will change the behaviour of how `stop_on_slope` works.

Currently, if the user wants the `KinematicBody` to not slide down a slope, the user calls `move_and_slide()` or `move_and_slide_with_snap()` and specifies the direction of the `floor normal` (normally up) and sets `stop_on_slope` to true. The user can also specify the `floor_max_angle` which defaults to 45°, which puts a limit on the angle considered a floor and hence the limit for the `stop_on_slope`. If the current `linear_velocity` passed is in the opposite direction to the `floor_normal` (i.e. normally directly down), the `KinematicBody` will not slide and the velocity returned will be zero. This zero velocity is normally feedback as the `linear_velocity` to the next call to `move_and_slide()` or `move_and_slide_with_snap()`, so the body stops.

If the current velocity is not in the opposite direction to the `floor_normal` (i.e. not just "falling" but moving) the body will slide and the velocity returned will be the updated velocity perpendicular to the collision normal.

I could be wrong, because I haven't tested it, but it looks like, this change will consider any movement up to the `floor_max_angle` (i.e. if it's on a floor) will now be stopped and a zero velocity returned if `stop_on_slope` is true. This would effectively be similar behaviour to `move_and_collide` (although sliding would still occur on walls and ceilings).

**Comment by AndreaCatania** [comment]
This change improves the old check that was good enough since in the previous integration was supposed to always send perfectly horizontal forces.

Now that it is reverted this check require too much precise velocities to make the player to stop on slopes so with my change any velocity that points in the opposite `up_direction` that falls under the threshold make the player to stop on slope.

So the nor the behaviour nor the concept are changed but just how the check is done.

**Comment by nonunknown** [comment]
hi, i've had a lot of headache with move_and_slide, because my game has irregular ground, and really stop_on_slop didnt make any change to kinematic3D's behavior, hope to see this change in action soon 
tkx @AndreaCatania

**Comment by madmiraal** [comment]
I have now tested both the 2D and 3D versions using a simple controller, and it does change the current behaviour:
[MoveandSlideTest.zip](https://github.com/godotengine/godot/files/4050665/MoveandSlideTest.zip)

Changing it to use a smaller, fixed value (COS_DOWN_DIRECTION_THRESHOLD) instead of `p_floor_direction` as in bd48323, does make it less noticeable, but it still changes the current behaviour. Furthermore, although 10° may work for the particular case you're testing against, it may not be enough for some cases and too much for others. At least when using `p_floor_direction`, the user would have been able to control the value.

Is the issue you're trying to fix the same as #34117? If so, have you considered using the suggestions provided? If not, please provide a sample project.

To be clear, I understand the frustrations of `stop_on_slope()` not working as expected, because the slope generates a lateral velocity that is typically passed back into the `move_and_slide()` or `move_and_slide_with_snap()` functions on the next `_process()` call. There is also no shortage of users raising [issues](https://github.com/godotengine/godot/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+move_and_slide). However, users can already achieve the same effect, by clamping the lateral velocity to zero when required. Alternatively, users can use a [`Rayshape`](https://docs.godotengine.org/en/3.1/classes/class_rayshape.html) as described [here](https://godotengine.org/article/godot-31-will-get-many-improvements-kinematicbody) when the current behaviour was first implemented.

In short, I don't think changing Godot's behaviour to fix for a specific use-case is the solution. As @akien-mga has [stated previously](https://github.com/godotengine/godot/issues/30029#issuecomment-504990050), the API needs to be refactored, but this requires the interface and its desired behaviour to be agreed first.

**Comment by AndreaCatania** [comment]
Can you please explain what "behaviour change" mean for you? because I don't really see your point.

The previous check, (which I've created) is meaning less and it's not the proper way of doing it. With this PR I'm just improving my old code.

> Furthermore, although 10° may work for the particular ...

The velocity is supposed to be always parallel to the ground. So the 10 deg mean that the player can climb up to 80 degrees slopes. It's quite unrealistic that someone want to climb such angles so it's not a number that fits only my use cases.

>  users can already achieve the same effect, by clamping the lateral velocity to zero when required. Alternatively, users can use a Rayshape as described here when the current behaviour was first implemented.

The idea behind the addition of `stop_on_slope` was to allow the user to avoid on slope player sliding with just one click. Even before was possible to stop on slope with complex workarounds!

> In short, I don't think changing Godot's behaviour to fix for a specific use-case is the solution. 

Well, this is not at all what it's happening here, and I can argue that following your reasoning any change is a behaviour change. Even a bug fix is a behaviour change :D ... But here we are off topic.

**Comment by AndreaCatania** [comment]
By the way, the proposed solution which is written here: https://github.com/godotengine/godot/issues/34117#issuecomment-562841402 is a workaround not the proper way of using the function.

**Comment by nonunknown** [comment]
@AndreaCatania I dont know if godot 3.2 beta 5 already integrated your changes to source, but i mean by behaviour changing is this:

actual code:
`velocity = move_and_slide(velocity, Vector3(0,1,0),true,1,1.0472,false)`
 as you can see stop_on_slope is true, but the kinematic body stays sliding:
https://imgur.com/J9t8T5K

**Comment by madmiraal** [comment]
> Can you please explain what "behaviour change" mean for you? because I don't really see your point.

By behaviour change, I mean a change from the [documented behaviour](https://docs.godotengine.org/en/3.1/classes/class_kinematicbody.html#class-kinematicbody-method-move-and-slide) of the function.

Unfortunately, the current documented behaviour of the `stop_on_slope` parameter is open to interpretation: "If `stop_on_slope` is `true`, body will not slide on slopes if you include gravity in `linear_velocity`."

Assuming that "include gravity" means have a negative `y` component in `linear_velocity` when the `floor_normal` equals `Vector.UP`, taken literally, this definition would result in the `KinematicBody` stopping on all surfaces that are considered slopes i.e. be the same as `move_and_collide()`. This is probably not the intended behaviour. The original PR (with [`f7d98c1`](https://github.com/godotengine/godot/commit/f7d98c1de2a365d6ebc976f81191b8e7f7c116c9)) would have resulted in this behaviour.

Instead, the current interpretation (assuming `floor_normal` equals `Vector.UP`) checks if the `linear_velocity` only has a negative `y` component, or more generally, if the `linear_velocity` and `floor_normal` vectors passed by the user are in opposite directions. It achieves this by checking if the combination of the normal (or direction) of the`linear_velocity` and `floor_normal` vectors passed by the user, has a length of less than a small value:
``` (body_velocity_normal + p_up_direction).length() < 0.01 ```
where `body_velocity_normal` equals `linear_velocity.normalized()` and `p_up_direction` equals `floor_normal`.

The result is, if `floor_normal` equals `Vector.UP`, the `linear_velocity` must have a negative `y` component and a `x` value (or `sqrt(x² + z²)` in the 3D case) of less than 0.01 times the length of the `y` component, which, to all intents and purposes requires the `x` and `z` components to be zero.

> The velocity is supposed to be always parallel to the ground. So the 10 deg mean that the player can climb up to 80 degrees slopes. It's quite unrealistic that someone want to climb such angles so it's not a number that fits only my use cases.

This PR currently (with  bd48323) proposes changing the above check to:
``` (body_velocity_normal.dot(p_up_direction) < COS_DOWN_DIRECTION_THRESHOLD ```
where COS_DOWN_DIRECTION_THRESHOLD is set to -0.98480774002 derived from -cos(10°). The dot product of two vectors **a** and **b** is [defined](https://en.wikipedia.org/wiki/Dot_product) geometrically as ||**a**|| ||**b**|| cos(_θ_), where _θ_ is the angle between the two vectors and the magnitude of a vector **a** is denoted ||**a**||. Since both vectors (assuming `floor_normal` is Vector.UP) have a magnitude of 1, this simply checks that the angle between the two vectors is between 170° and 190° (-cos(10°) = cos(170°) = cos(190°)) i.e. roughly pointing in opposite directions. This effectively increases the tolerance of the size of the `x` value (or `sqrt(x² + z²)` in the 3D case) to less than approximately 0.17 times the `y` component.

> > In short, I don't think changing Godot's behaviour to fix for a specific use-case is the solution.

>Well, this is not at all what it's happening here

Since `move_and_slide()` uses
``` linear_velocity - floor_normal * linear_velocity.dot(floor_normal) ```
to calculate the returned velocity, the `x` component (or `sqrt(x² + z²)` in the 3D case) returned will be sin(_α_) times the size of the `y` component, where _α_ is the angle of the slope. For slope angles below the 10° currently defined by this PR, the `x` component (or `sqrt(x² + z²)` in the 3D case) will be less than the required 0.17 times the `y` component, for slope angles above 10°, the user's controller will still need to reduce this `x` value (or `sqrt(x² + z²)` in the 3D case) to less than 0.17 times the `y` component to stop the `KinematicBody` from sliding.

Conversely, with this PR, other controllers, which currently clamp the `x` (and `z` components) to zero when required, will now stick on flat surfaces for velocities with an `x` value (or `sqrt(x² + z²)` in the 3D case) of less than 0.17 times the `y`component instead being able to smoothly accelerate by slowly increasing the `x` component.


In short this PR provides a hard-coded snapping function, which may not be sufficient for some users' controllers and may be undesired for other users' controllers.

I agree that the `stop_on_slope` doesn't work as desired, but I don't agree that this PR is the solution. It will fix specific use cases, but not all, and it will create new issues. The API needs to be fixed, but, as I stated previously, the interface and its desired behaviour should be agreed first.

**Comment by AndreaCatania** [comment]
Do we agree that to move a character we need a velocity parallel to the ground?

**Comment by AndreaCatania** [comment]
> @AndreaCatania I dont know if godot 3.2 beta 5 already integrated your changes to source, but i mean by behaviour changing is this:
> 
> `velocity = move_and_slide(velocity, Vector3(0,1,0),true,1,1.0472,false)`
> as you can see stop_on_slope is true, but the kinematic body stays sliding:
> 
> https://imgur.com/J9t8T5K

The PR is not merged yet, however is difficult tell what's going on without seeing the code

**Comment by madmiraal** [comment]
> @madmiraal Do we agree that to move a character we need a velocity parallel to the ground?

Yes.

To be specific, the `move_and_slide()` and `move_and_slide_with_snap()` functions return a modified `linear_velocity` vector that is perpendicular to the last collision point's normal. If this collision point is the "ground" then it will be parallel to it.

**Comment by AndreaCatania** [comment]
@madmiraal Considering this check`(body_velocity_normal + p_up_direction).length() < 0.01` do we agree that `0.01` is not an angle but rather a meaningless distance and there are more meaningful way to check the velocity direction in respect to the `up_direction` that can improve future maintanance?

**Comment by madmiraal** [comment]
> @madmiraal Considering this check `(body_velocity_normal + p_up_direction).length() < 0.01` do we agree that `0.01` is not an angle but rather a meaningless distance and there are more meaningful way to check the velocity direction in respect to the `up_direction` that can improve future maintanance?

`0.01` is simply a small value to allow for the numerical imprecision of floats. It could probably have been set to `UNIT_EPSILON`.

`(body_velocity_normal + p_up_direction).length() < 0.01` checks to see whether the direction of the `linear_velocity` vector and the `up_direction` vector (both passed by the user) are in opposite directions. Note: It assumes the users passes a normalised vector as the `up_direction` parameter.

If the user passes `Vector.UP` as the `up_direction` parameter, this provides a simple check that the `y` component of `linear_velocity` is negative and the `x` and `z` components are zero, while allowing for scenarios when the user chooses another direction for up.

In turn, it provides a simple way for the user to stop the `KinematicBody` from sliding when using `move_and_slide()` or `move_and_slide_with_snap()` and `stop_on_slide` is set to `true`: Simply make the `x` and `z` components zero.

The problem arises when the user uses a controller that doesn't set the 'x' and 'z' components to zero when there is no input. Rather the 'x' and 'z' components are moved incrementally towards zero. This is done to provide a smooth deceleration. However, this approach fails when combining it with `move_and_slide()` or `move_and_slide_with_snap()` and the increment towards zero is less than the change generated by the slide. This problem is difficult for the user to solve, because the change generated by the slide is dependent on the steepness of the slope, and many are simply copying another controller and expecting it to work in all scenarios.

Therefore, this is not a problem with the slide functionality of the `move_and_slide()` and `move_and_slide_with_snap()` functions. They are doing what they are supposed to. The problem is the user doesn't understand why on some slopes their character slides down.

The solution is to ensure that the `x` and `z` components are snapped to zero when required, but the user should specify when this is required. This allows the user to fine-tune it to her requirements. The `move_and_slide()` and `move_and_slide_with_snap()` functions should not be specifying when this is required or even that it is required.

Remember, the `move_and_slide()` and `move_and_slide_with_snap()` functions are convenience functions. The user doesn't need to use them. All the functionality can be done by the user using `move_and_collide()` and updating their character's velocity using the [`KinematicCollision`](https://docs.godotengine.org/en/3.1/classes/class_kinematiccollision.html) or [`KinematicCollision2D`](https://docs.godotengine.org/en/3.1/classes/class_kinematiccollision2d.html) object returned. This is after all what `move_and_slide()` is doing, and what `move_and_slide_with_snap()` is doing by calling `move_and_slide()`.

The reason they are provided as convenience functions is that they are both common requirements, and, move importantly, it is difficult to calculate the updated velocity vector from the collision normal to generate a slide. `linear_velocity - floor_normal * linear_velocity.dot(floor_normal)` is not intuitive, but it works and it also automatically reduces the velocity proportional to the steepness of the slope albeit not in a correct physics way.

In short, I don't agree that `0.01` is "a meaningless distance" or that there is a need for a "more meaningful way to check the velocity direction". I do agree there is a need to update the existing code to "improve future _maintenance_".

**Comment by AndreaCatania** [comment]
> In short, I don't agree that 0.01 is "a meaningless distance" or that there is a need for a "more meaningful way to check the velocity direction". I do agree there is a need to update the existing code to "improve future maintenance".

Despite the fact that if you don't agree on my sentences you should not say that we need to update the existing code... If it's fine, why it should be ever improved? But never mind, let's follow my reasoning and arrive to a final point.

As it's possible to read in your comments the `move_and_slide` function accept a velocity and moves toward that direction.

To define if the player is on slope or on wall the function checks the orientation between the `collision_normal` and the `up_direction`. Instead, the slide on slope feature doesn't check the direction and I think that using a check that has a threshold in radians, rather than the current check that only checks it's perpendicularity with a meaningless threshold, is a good way of unifying the logic of the function.

Simply, and despite maintenance improvement, the benefit is that for the user will be much easier use this function if he knows that any velocity that is toward down and falls inside the range of __-10°/+10°__ is considered a velocity that allows the player to stop on slopes.

But this approach seems wrong for you. However, seems that improve it is the right thing to do even for you, so what do you propose?

**Comment by madmiraal** [comment]
> what do you propose?

The API needs to be refactored, but this requires the interface and its desired behaviour to be agreed first.

In the meantime, we need to help users understand that for `stop_on_slope` to work, they need to ensure that their controller sets the `x` and `z` components of the `linear_velocity` vector to zero when required.

**Comment by akien-mga** [comment]
The important thing now is to ensure that people porting their projects over from 3.1 won't have a big compatibility breakage. I don't know what's the current state but with many bug fixes since 3.1 I don't know where we stand, and whether this PR makes things closer to the general expectation or not.

The KinematicBody API has grown organically to something quite messy, and the scope right now should not be to refactor it or make it even more complex - IMO a full refactor should happen in 4.0, breaking compatibility to make the API clearer, cleaner and better.

For 3.2 (i.e. right now, we're at RC stage), I want the minimum amount of disruption for existing projects, while allowing as many use cases as possible (and if it implies using a few lines of code instead of a single call, that's fine).

**Comment by AndreaCatania** [comment]
I don't think that this PR changes how the function works, rather it establish a more understandable way of how it can be used:

_Any velocity toward down direction, that falls in the range of -10° +10° are now considered stopping velocity._

I've already explained why -10° and +10° are enough but If you still think that this is too much behavioural change I'm fine to close it.

**Comment by Paavs-Git** [comment]
The [documentation](https://docs.godotengine.org/en/latest/tutorials/physics/physics_introduction.html) describes 'move_and_slide' as 'a convenient way to implement sliding movement without writing much code.';
Currently this is hardly case for anything other than very basic movement, the workaround posted [here](https://github.com/godotengine/godot/issues/34117#issuecomment-562841402) is not only not convenient, it doesn't work by itself, one must also do something like:
`velocity.y = move_and_slide(velocity, Vector3(0, 1, 0), true).y`
This looks very hacky to me.
It seems impossible to implement good sliding and slope behaviour with the current move_and_slide functions without these sort of odd workarounds.
To me this PR makes a lot of sense as a good step towards making move_and_slide convenient and easier to use.

**Comment by 0xc0c0a1** [comment]
Reading this thread be like
image

**Comment by pouleyKetchoupp** [comment]
@AndreaCatania Stop on slope not working in some cases might be fixed with #51904 (since it fixes #34117), but I'm not sure about your specific use case. Would you have a MRP that I can test to confirm?

**Comment by filipworksdev** [comment]
I have tested this in 3D in 3.x recently and I didn't see any major difference between with and without the pr. Is it possible this may have already been fixed?

**Comment by akien-mga** [comment]
There have been a lot of fixes to `move_and_slide` since then indeed, included in 3.4.

If there are still problematic edges cases, the best would be to open a bug report with a minimal reproduction project to assess them against the current API.
